File: test-utils/src/main/java/io/appmetrica/analytics/testutils/MockedConstructionRule.java
package io.appmetrica.analytics.testutils;

import org.junit.rules.ExternalResource;
import org.mockito.MockedConstruction;
import org.mockito.Mockito;

public class MockedConstructionRule<T> extends ExternalResource {

    private MockedConstruction<T> mock;

    private final Class<T> clazz;
    private final ConstructionArgumentCaptor<T> interceptor;

    public MockedConstructionRule(Class<T> clazz) {
        this(clazz, null);
    }

    public MockedConstructionRule(Class<T> clazz, MockedConstruction.MockInitializer<T> initializer) {
        this.clazz = clazz;
        if (initializer == null) {
            interceptor = new ConstructionArgumentCaptor<>();
        } else {
            interceptor = new ConstructionArgumentCaptor<>(initializer);
        }
    }

    public MockedConstruction<T> getConstructionMock() {
        return mock;
    }

    public ConstructionArgumentCaptor<T> getArgumentInterceptor() {
        return interceptor;
    }

    @Override
    protected void before() throws Throwable {
        super.before();
        mock = Mockito.mockConstruction(clazz, interceptor);
    }

    @Override
    protected void after() {
        super.after();
        mock.close();
    }

}

File: test-utils/src/main/java/io/appmetrica/analytics/testutils/MockedConstructionRule.kt
package io.appmetrica.analytics.testutils

import org.junit.rules.ExternalResource
import org.mockito.MockedConstruction
import org.mockito.Mockito
import org.mockito.kotlin.KStubbing
import org.mockito.kotlin.stubbing
import kotlin.reflect.KProperty

class MockedConstructionRule<T>(
    private val clazz: Class<T>,
    initializer: MockedConstruction.MockInitializer<T>? = null
) : ExternalResource() {
    constructor(clazz: Class<T>) : this(clazz, null)

    lateinit var constructionMock: MockedConstruction<T>
        private set
    val argumentInterceptor: ConstructionArgumentCaptor<T> = ConstructionArgumentCaptor(initializer)

    @Throws(Throwable::class)
    override fun before() {
        super.before()
        constructionMock = Mockito.mockConstruction(clazz, argumentInterceptor)
    }

    override fun after() {
        super.after()
        if (this::constructionMock.isInitialized) {
            constructionMock.close()
        }
    }

    operator fun <C> getValue(thisRef: C, property: KProperty<*>): T {
        return constructionMock.constructed().single()
    }
}

operator fun <C, T> ConstructionArgumentCaptor<T>.getValue(thisRef: C, property: KProperty<*>): List<Any> {
    return arguments.single()
}

inline fun <reified T> constructionRule(noinline initializer: KStubbing<T>.(T) -> Unit = {}) =
    MockedConstructionRule(T::class.java) { mock, _ -> stubbing(mock, initializer) }
